sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +301,183 @@ def reduce(self, word, exclude=None):
+
+        # Keep track of the accept_states.
+        accept_states = []
+
+        # The symbols present in the new rules are the symbols to be verified at each state.
+        # computes the automaton_alphabet, as the transitions solely depend upon the new states.
+        for rule in all_rules:
+            automaton_alphabet += rule.letter_form_elm
+        automaton_alphabet = set(automaton_alphabet)
+
+        # Compute the proper prefixes for every rule.
+        for r in all_rules:
+            proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            proper_prefixes[r] = letter_word_array
+
+        # Add dead states.
+        # This is essentially checking the left hand side of the rules.
+        for rule in all_rules:
+            if not rule in self.reduction_automaton.states:
+                self.reduction_automaton.add_state(rule, state_type='d', rh_rule=all_rules[rule])
+
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in self.reduction_automaton.states:
+                    self.reduction_automaton.add_state(elem, state_type='a')
+                    accept_states.append(elem)
+
+        # Add new transitions for every state.
+        for state in self.reduction_automaton.states:
+            current_state_name = state
+            current_state_type = self.reduction_automaton.states[state].state_type
+            # Transitions will be modified only when current_state belongs to the proper_perfixes of the new rules.
+            # The rest are ignored if they cannot lead to a dead state after a finite number of transisitons.
+            if current_state_type == 's':
+                for letter in automaton_alphabet:
+                    if letter in self.reduction_automaton.states:
+                        self.reduction_automaton.states[state].add_transition(letter, letter)
+                    else:
+                        self.reduction_automaton.states[state].add_transition(letter, current_state_name)
+            else:
+                # Store subwords of current state in an array.
+                current_state_name_array = []
+                next_sub = current_state_name
+                while len(next_sub):
+                    current_state_name_array.append(next_sub)
+                    next_sub = next_sub.subword(1, len(next_sub))
+                # Check if the transition to any new state in posible.
+                for sub_word in current_state_name_array:
+                    if sub_word in accept_states:
+                        # Add accept states.
+                        if current_state_type == 'a':
+                            for letter in automaton_alphabet:
+                                next = current_state_name*letter
```

Comment:
`next` is a python function, it doesn't seem to cause problems but I'd still rename it to e.g. `_next`